src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/factory/generator/flow/MultiTableIngressYRuleGenerator.java

```diff
@@ -0,0 +1,116 @@
+
+@SuperBuilder
+public class MultiTableIngressYRuleGenerator extends MultiTableIngressRuleGenerator {
+
+    protected final MeterId sharedMeterId;
+
+    @Override
+    public List<SpeakerCommandData> generateCommands(Switch sw) {
+        if (flow.isOneSwitchFlow()) {
+            throw new IllegalStateException("Y-Flow rules can't be created for one switch flow");
+        }
+        List<SpeakerCommandData> result = new ArrayList<>();
+        FlowEndpoint ingressEndpoint = FlowSideAdapter.makeIngressAdapter(flow, flowPath).getEndpoint();
+        if (!ingressEndpoint.getSwitchId().equals(sw.getSwitchId())) {
+            throw new IllegalArgumentException(format("Path %s has ingress endpoint %s with switchId %s. But switchId "
+                            + "must be equal to target switchId %s", flowPath.getPathId(), ingressEndpoint,
+                    ingressEndpoint.getSwitchId(), sw.getSwitchId()));
+        }
+        FlowSpeakerCommandData command = buildFlowIngressCommand(sw, ingressEndpoint);
+        if (command == null) {
+            return Collections.emptyList();
+        }
+        result.add(command);
+
+        // TODO(tdurakov): since it's shared meter, this build might be moved outside.
+        SpeakerCommandData meterCommand = buildMeter(flowPath, config, sharedMeterId, sw);
+        if (meterCommand != null) {
+            result.add(meterCommand);
+            command.getDependsOn().add(meterCommand.getUuid());
+        }
+
+        return result;
+    }
+
+
+
+    private FlowSpeakerCommandData buildFlowIngressCommand(Switch sw, FlowEndpoint ingressEndpoint) {
+        List<Action> actions = new ArrayList<>(buildTransformActions(
+                ingressEndpoint.getInnerVlanId(), sw.getFeatures()));
+        actions.add(new PortOutAction(new PortNumber(getOutPort(flowPath, flow))));
+
+        FlowSpeakerCommandDataBuilder<?, ?> builder = FlowSpeakerCommandData.builder()
+                .switchId(ingressEndpoint.getSwitchId())
+                .ofVersion(OfVersion.of(sw.getOfVersion()))
+                .cookie(flowPath.getCookie().toBuilder().yFlow(true).build())
+                .table(OfTable.INGRESS)
+                .priority(isFullPortEndpoint(ingressEndpoint) ? Constants.Priority.Y_DEFAULT_FLOW_PRIORITY
+                        : Constants.Priority.Y_FLOW_PRIORITY)
+                .match(buildIngressMatch(ingressEndpoint))
+                .instructions(buildInstructions(sw, actions));
+
+        if (sw.getFeatures().contains(SwitchFeature.RESET_COUNTS_FLAG)) {
+            builder.flags(Sets.newHashSet(OfFlowFlag.RESET_COUNTERS));
+        }
+        return builder.build();
+    }
+
+    private Instructions buildInstructions(Switch sw, List<Action> actions) {
+        Instructions instructions = Instructions.builder()
+                .applyActions(actions)
+                .goToTable(OfTable.POST_INGRESS)
+                .build();
+        addMeterToInstructions(sharedMeterId, sw, instructions);
```

Comment:
Probably missing meter availability check before call `addMeterToInstructions`.

Reply:
it's part of the method